### ISSUE TITLE
[Bug] Upgrade to 3.3.2 Fails, SSL_CREDENTIALS_TYPE=KEYSTORE does not work anymore

### ISSUE THREAD

**Issue by KenSense** [thingsboard/thingsboard#5617]
**Describe the bug**
Upgrade to 3.3.2PE from 3.3.1PE failed
_Unable to start web server; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'tomcatServletWebServerFactory' defined in class path resource [org/springframework/boot/autoconfigure/web/servlet/ServletWebServerFactoryConfiguration$EmbeddedTomcat.class]: Initialization of bean failed; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'sslCredentialsWebServerCustomizer': Unsatisfied dependency expressed through field 'httpServerSslCredentialsConfig'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'httpServerSslCredentials': Invocation of init method failed; nested exception is java.lang.RuntimeException: HTTP Server SSL Credentials: Failed to init SSL credentials configuration.
ThingsBoard upgrade failed!_

I changed the new defaults from PEM to (MQTT_)SSL_CREDENTIALS_TYPE=KEYSTORE so it would work as before but it still fails.
SSL_KEY_STORE_TYPE=PKCS12 stayed the same

**Your Server Environment**
* own setup
  * local infrastructure
  * ThingsBoard PE 3.3.1
  * OS: Ubuntu 20.04LTS

* Connectivity
  * MQTT with TLS
  * HTTPS

**To Reproduce**
Upgrade PE system with 'KEYSTORE ' (P12) certificates

**Comment by ashvayka** [comment]
The default change KEYSTORE -> PEM supports (at least it should) backward compatibility. Because if pem file is missing the system will look for jks. 

please attach part of thingsboard.yml or thingsboard.conf that is related to this topic.

**Comment by ashvayka** [comment]
Also please add the yml part for the HTTP (not MQTT). And send the logs to support@thingsboard.io. We will analyze them and update this thread with the root cause.

**Comment by KenSense** [comment]
I saw that in the log 'PEM didn't work trying keystore', hense changing the .yml to keystore didn't help.

I tried to use PEM files instead but also didn't work.
PEM:
Caused by: java.lang.IllegalArgumentException: Unable to load private key neither from certFile:
or KEYSTORE:
Caused by: java.lang.IllegalArgumentException: Failed to get private key from the keystore or pem files. Please check if the private key exists in the keystore or pem files and if the provided private key password is valid.

I tried different formats, converted with this command, the .P12 is ok because that was working before.
openssl pkcs12 -in keys.p12 -out keys.p12.pem

Will send email with config files with the same subject as this issue

**Comment by KenSense** [comment]
Would the update complete normally if I first disable TLS and then try to enable it again?

**Comment by ashvayka** [comment]
The error from your log is quite clear:

```text
Failed to get private key from the keystore or pem files. 
Please check if the private key exists in the keystore or pem files and if the provided private key password is valid.
```

Looks like you don't have the private key file in your pem/jks. This is the root cause. Most likely it is not caused by the update of the version.

**Comment by KenSense** [comment]
> The error from your log is quite clear:
> 
> ```
> Failed to get private key from the keystore or pem files. 
> Please check if the private key exists in the keystore or pem files and if the provided private key password is valid.
> ```
> 
> Looks like you don't have the private key file in your pem/jks. This is the root cause. Most likely it is not caused by the update of the version.

Well I didn't change the files or password settings and they worked fine in 3.3.1 and I checked they are ok.
I disabled TLS in .yml this allowed me to complete the update, but enabling it again to start the service obviously fails again with the same "Failed to get private key from the keystore" error.

So what did change, I didn't change anything?

**Comment by ashvayka** [comment]
we have not received the email with config files yet. So it is hard to guess. Also make sure that the file names in the yml/conf match the file name of your file.

**Comment by KenSense** [comment]
> 
Send yesterday at 16:54  to support@thingsboard.io

I only disabled TLS for HTTPS now and started the service, it seems MQTT on 8883 is working.
Trying to install HAproxy now to take over the TLS job.

**Comment by KenSense** [comment]
It's back online now, SSL_ENABLED=false, and enable it in HAproxy.

You still have a problem but it's not longer used by me.

**Comment by ashvayka** [comment]
We have not received your email. Please identify the email address you have used to send an email for debug.

**Comment by KenSense** [comment]
> We have not received your email. Please identify the email address you have used to send an email for debug.

The subject line is the same as the one of this issue, filter in that, check your spam.
The github link is in the body (I didn't get a failed delivery so it did end up in a mailbox)

**Comment by ashvayka** [comment]
I have checked everything and there is no such email. Unfortunately, i will need to close this ticket if we can't get the logs to investigate.

**Comment by KenSense** [comment]
For a previous problem I remember you also didn't receive the email.
maybe you should give a direct email address because this one isn't working.

About closing the issue, have you tested it? (using an P12 file with password)
I'm using a workaround now but that doesn't mean your bug is solved.

# TLS settings
export HTTP_BIND_PORT=8443
export SSL_ENABLED=true
export SSL_CREDENTIALS_TYPE=KEYSTORE
export SSL_KEY_STORE=classpath:keystore/Cert_2021-06-10_incl_tree_noroot.p12
export SSL_KEY_STORE_PASSWORD=*********
export SSL_KEY_STORE_TYPE=PKCS12
export SSL_KEY_ALIAS=te-13eb49d0-253c-4120-9b04-15b6eb7bf412

**Comment by ashvayka** [comment]
First of all, the "SSL_*" settings you have attached is introduced in 3.3.2 (I am talking about SSL_*, not MQTT_*). So, the specific settings you have attached can't cause the issue during the upgrade. They were not present in 3.3.1. 
Second, the "classpath:" settings are something you should avoid. See the official docs about how to set up the [HTTPS](https://thingsboard.io/docs/user-guide/ssl/http-over-ssl/) and [MQTTS](https://thingsboard.io/docs/user-guide/mqtt-over-ssl/). It has important notice:

```
Make sure the certificate files are reachable by ThingsBoard process:

Linux: use /etc/thingsboard/conf folder. Make sure the files have the same permissions as thingsboard.conf; Use relative file path, e.g. server.pem;
Docker Compose: mount or use existing volume to /config folder of the container; Use full file path, e.g. /config/server.pem;
K8S: mount separate volume to /https-config or similar folder. Use full file path, e.g. /https-config/server.pem;
Windows: use C:\Program Files (x86)\thingsboard\conf\ folder. Make sure the files have the same permissions as thingsboard.conf; Use relative file path, e.g. server.pem;
```

Third, yes, we test for backward compatibility during updates. We intentionally set default credentials type to PEM since it is more common for new users and we also added a fallback to JKS if the *.pem file is missing.

Forth, our mail server is provided by G-suite, it is nearly perfect and we have minimal issues with it. I suspect either your email server blocks some specific content or you type the email address instead of copy-pasting it.

**Comment by KenSense** [comment]
I did set SSL_ENABLED to false so the update to 3.3.2 would not fail anymore.
So my further SSL tests that I did were on 3.2.2 installed successfully.
Enabled SSL again and I get the same error to start the service as I got during the upgrade process.
Indeed not MQTT since that P12 (in the same folder) kept working strangely because the same PEM code change was done there.)

SSL settings that are new in 3.2.2 are SSL_CREDENTIALS_TYPE and SSL_PEM_CERT  etc. not the SSL_KEY_STORE*** settings I listed.

I tested with and without 'classpath' since I wasn't sure it was needed, it didn't change, I also moved it to another directory but the effect was the same, it could access it in 3.3.1 so I didn't expect it to change.
